wp-job-manager-functions.php

```diff
@@ -1712,3 +1712,26 @@ function job_manager_get_salary_unit_options( $include_empty = true ) {
 	 */
 	return apply_filters( 'job_manager_get_salary_unit_options', $options, $include_empty );
 }
+
+/**
+ * Delete filled_jobs_transient and expired_jobs_transient transients when a job listing is saved or updated.
+ *
+ * @param int     $post_id  The ID of the post being saved.
+ * @param WP_Post $post The post object being saved.
+ */
+function delete_job_listings_transients_on_save( int $post_id, WP_Post $post ): void {
+
+	if ( 'job_listing' === $post->post_type ) {
+
+		if ( '1' === get_post_meta( $post_id, '_filled', true ) ) {
+			delete_transient( 'hide_filled_jobs_transient' );
+		}
+
+		$post_status = [ 'publish', 'expired', 'trash' ];
+		if ( in_array( $post->post_status, $post_status, true ) ) {
+			delete_transient( 'hide_expired_jobs_transient' );
+		}
+	}
+
+}
+add_action( 'save_post', 'delete_job_listings_transients_on_save', 10, 2 );
```

Comment:
These transients need to be cleared whenever:

- A post is marked as `Filled`
- A post is changed from `publish` to `expired` or `trash` post status

I'm uncertain about this implementation and am quite open to feedback about improvements or changes to this implementation.

It's just struck me that the `filed` transient also needs to be cleared if a job is set from `filled` to not filled, but I'll wait on review before making any further changes in case this isn't the right direction as a whole.

Reply:
The approach makes sense to me. I'm a bit tempted to add/remove the post ID directly to the cache instead of re-running the query after every change. That could be faster but a bit more risky.

[Looks like the meta is set with `update_post_meta`](https://github.com/Automattic/WP-Job-Manager/blob/38019efba98ce55b2c881f30003a3b8059f9e1e1/includes/class-wp-job-manager-shortcodes.php#L214), so the `save_post` hook won't be enough and we should hook into the `updated_post_meta` action.

Also found this class that could be a good home for some of the caching:
[WP_Job_Manager_Cache_Helper](https://github.com/Automattic/WP-Job-Manager/blob/edfd346146b624b13f09b6a8c1e3881b98a40187/includes/class-wp-job-manager-cache-helper.php)

And I don't know much about it, but could using the object cache via [`wp_cache_add/delete`](https://developer.wordpress.org/reference/functions/wp_cache_add/) be a better fit? Though looks like the cache helper above also uses transients.

Reply:
After a further look into the class and some discussion with Jake, I'm going to keep to the current transient implementation. It doesn't look like there's anything too useful in the class for the caching that we want to do.

Reply:
Yeah updating just the affected ID in the transient array is probably too complicated to be worth it.